src/Lecture2.hs

```diff
@@ -260,8 +313,23 @@ data EvalError
 {- | Having all this set up, we can finally implement an evaluation function.
 It returns either a successful evaluation result or an error.
 -}
+containsT :: [(String, Int)] -> String -> Bool
+containsT [] _ = False
+containsT [x] a = fst x == a
+containsT (x:xs) a = containsT [x] a || containsT xs a
+
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit x) = Right x
+eval vars (Add a b) = case (r1, r2) of
+  (Left err, _) -> Left err
+  (_, Left err) -> Left err
+  (Right result1, Right result2) -> Right $ result1 + result2
```

Comment:
That's a great pattern matching handling! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️